debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java

```diff
@@ -199,7 +199,7 @@ protected void doEnqueue(T record) throws InterruptedException {
         synchronized (this) {
             while (queue.size() >= maxQueueSize || (maxQueueSizeInBytes > 0 && currentQueueSizeInBytes >= maxQueueSizeInBytes)) {
                 // notify poll() to drain queue
-                this.notify();
+                this.notifyAll();
```

Comment:
Worth talking to @krnaveen14 about this, who did this change originally. Note it used to be even `notifyAll()` [initially](https://github.com/debezium/debezium/commit/62045f6430796bc0d3bee69ba1a8ba1ea5abcf1d). Tbh. I'd prefer to use any of the more modern synchronisation primitives, I haven't seen wait/notify being used out in the wild for ages.

Reply:
@gunnarmorling Yes, we need two locks for that but I'd like to get a fix out first and we can rewrite the code later.

Reply:
@gunnarmorling @jpechane FYI, It was initially defined as `notifyAll()`, but was suggested to use `notify()` during PR Review as it's not a Best Practice to use the former. Also since my knowledge with PostgreSQL CDC (pgoutput plugin) is that it's doing sequential WAL decode and so assumed that there would be only one writer.

I can strongly confirm that deadlock / locking issues would arise if there is more than one writer thread calling the same **Instance** of `ChangeEventQueue`. So it should be changed to `notifyAll()` and there shouldn't be any impact due to this.

Reply:
@jpechane Similar `notifyAll()` changes required in `poll()` method of `ChangeEventQueue`. Otherwise it would cause `deadlock` in different scenario.